Objects/typeobject.c

```diff
@@ -5341,49 +5399,107 @@ type_setattro(PyObject *self, PyObject *name, PyObject *value)
             name, type->tp_name);
         return -1;
     }
-    if (PyUnicode_Check(name)) {
-        if (PyUnicode_CheckExact(name)) {
-            Py_INCREF(name);
+    if (!PyUnicode_Check(name)) {
+        PyErr_Format(PyExc_TypeError,
+                     "attribute name must be string, not '%.200s'",
+                     Py_TYPE(name)->tp_name);
```

Comment:
Just checking my understanding: this was already enforced in `_PyObject_GenericSetAttrWithDict`, right? So it's not a new restriction, it's just that the error is raised earlier in the code path.